File: bin/pooltnt.py
#!/usr/bin/env python
#
import sys
import time
import multiprocessing 
import subprocess
import shlex
import os
import glob
mydir = ""
runpth = ""
def do_align(subdir):
    print "starting align for %d\n" % (subdir)
    tstdir = "%s%d" % (mydir,subdir)
    if not os.access(tstdir,os.F_OK):
        print tstdir, "does not exist\n"
        return 1
    tstdir += "/oid.tre"
    if os.access(tstdir,os.F_OK):
        print tstdir, "already done, skipping\n"
        return 0
    atstr = "%sorthologid.pl -at \"^%d$\"" % (runpth,subdir)
    
    args = []
    args = shlex.split(atstr)
    print "str: %s \n" % atstr
    print "args",args
    p = subprocess.Popen(args)
    sys.stdout.flush()
    rc = p.wait()
    print "end align %s rc %d\n" % (subdir,rc)
    sys.stdout.flush()
## end do_align
mindir = 0
maxdir = 0
if __name__ == '__main__':
    
    print  "time is %s cpu %d \n" % (time.ctime(),time.clock())
    nargs = len(sys.argv)
    if nargs>2:
        maxdir = int(sys.argv[2])
        mindir = int(sys.argv[1])
    if nargs<2 or maxdir<mindir:
        print "pooltnt.py min max\n"
        print "must have range of dir to run\n"
        exit(1)
    mydir = os.getcwd() + '/data/'
    ncpu = multiprocessing.cpu_count()
    runpth = os.environ["OID_HOME"] + '/bin/'
    print "$OID_BIN ",runpth
    print "we were called with %d cpus\n" % (ncpu)
    if nargs>3:
        ncpu = int(sys.argv[3])
        print "we were explicitly given %d cpus to use\n" % (ncpu)
#    dirlst = list(range(mindir,maxdir+1)) # set up list of processes to run
    dirlst = list(range(maxdir,mindir-1,-1)) # set up list of processes to run
# highest is smallest and fastest
    p = multiprocessing.Pool(ncpu)
    poolrc = p.map(do_align,dirlst)
    p.close()
    p.join()
    prstr = "time is %s cpu %d end \n" % (time.ctime(),time.clock())
    print "pool: ",prstr
    ou = open ("log/job/pooldone",'w')
    print >>ou, "min %d, max %d\n" % (mindir,maxdir)
    print >>ou, prstr
    ou.close()
    sys.stdout.flush()
    

File: bin/pooltnt.pl
#!/usr/bin/env perl

use strict;
use Cwd;
use Time::Piece;

use Fcntl;
my $OID_HOME;
my $OID_USER_DIR;
my $OID_BLAST;
my $OID_BLASTDIR;
my $HPC;

BEGIN {
    $OID_HOME = $ENV{'OID_HOME'};
    die "Environment variable OID_HOME is not defined ... exiting.\n"
      if !defined($OID_HOME);
    $HPC          = $ENV{'HPC'};
    $HPC          = 'P' if !defined($HPC);
    $OID_USER_DIR = $ENV{'OID_USER_DIR'};
    die "Environment variable OID_USER_DIR is not defined ... exiting.\n"
      if !defined($OID_USER_DIR);
    $OID_BLAST = "$OID_USER_DIR/blast";
}

use lib "$OID_HOME/lib";
use OrthologID;
my @pidtbl  = ();
my @opentbl = ();

#my @tntcmd = qw(oid2.proc oid3.proc);
my @tntcmd = ();
my $nproc  = 2;
my $tntstr = 'pf';
my $nthr   = 2;
my $curthr = 0;
my $tm;
my $pooldone = 'log/job/pooldone';
my $savDir   = getcwd;
$tm = localtime;
print "pooltnt.pl in $savDir at $tm\n";

#    chdir "1";
#    die "runtntmx.pl nproc nthreads prefix\n" if (@ARGV<3);
die "pooltnt.pl startfam,endfam,ncpu\n" if ( @ARGV < 3 );
my $startfam = $ARGV[0];
my $endfam   = $ARGV[1];
$nthr = $ARGV[2];
print "we run from fam $startfam to $endfam on $nthr cpus\n";

#    foreach my $tnt (@tntcmd){
for ( my $fam = $startfam ; $fam <= $endfam ; $fam++ ) {
    my $pid;

    #        sleep 2;
    #        next if $pid = fork;  #parent
    if ( -s "data/$fam/oid.tre" ) {
        print "$fam already done\n";
        next;    #just get next fam
    }
    if ( $pid = fork() ) {    # parent
        $curthr++;
        next if ( $curthr < $nthr );
        my $child = wait();
        $tm = localtime;
        print "$tm $child completed\n";
        $curthr--;
        next;    #so we can fork
    }

    $tm = localtime;
    print "$tm starting $fam\n";
    if ( -s "data/$fam/oid.tre" ) {
        print "$fam already done\n";
        exit;
    }
    local $SIG{ALRM} = sub { print "$pid timed out $fam\n"; exit };
    my $famdol = '^' . "$fam" . '$';
    print "in child  fam $fam\n";
    alignFamily("$famdol");
    alarm 600;    #these are fast rtns but may hang
    makeTree("$famdol");
    alarm 0;      #cancel alarm it works

    #        my $status = system("tnt p $tnt 0</dev/null 1>&0 2>&0");
    #
    my $tm = localtime;
    print "$tm tnt with $fam done \n";
    exit;
}

#    1 while (wait() != -1);  ## does not seem to work - also does not record last few
while ( $curthr > 0 ) {
    my $child = wait();
    $tm = localtime;
    print "$tm $child completed\n";
    $curthr--;
}

open( FI, ">$pooldone" );
print FI "$startfam,$endfam\n";
close FI;
$tm = localtime;
print "$tm pooltnt done\n";

